### PR TITLE
Native driver for Eurobraille Esys, Esytime and newer Iris displays

### LINKED ISSUE #7488: Native driver for Eurobraille Esys displays
Formally a duplicate of closed #5176, but that one specifically talks about including the current add-on based driver by Eurobraille.

I wonder, has a native driver been considered, either by Eurobraille, NV Access or a third party? The current add-on based driver has the following issues:

1. Slow response for several keys, including routing and scrolling. This also occurs in JAWS and Supernova, probably an issue in a common part of the current driver
2. No native integration for the Esys braille keyboard, so NVDA's braille input does not work
3. When trying to initialize the driver without a display connected, it takes ages to time out. Even when a display is connected, there is a bug which causes the display not always being detected properly.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #7488.
Note that this also contains the commits from #7732, which is currently incubating. I'm sure git merging will deal with that properly.

### Summary of the issue:
For Eurobraille displays, users currently rely on an add-on packaged driver that has some drawbacks for NVDA users, including:

1. The inability to assign gestures for many keys using NVDA's input gesture framework
2. The use of a custom braille input layer
3. The add-on is quite huge to distribute and, due to its size, can't be integrated in NVDA core
4. The add-on based driver won't support braille display auto detection (#1271)

### Description of how this pull request fixes the issue:
This pr implements a native driver for Eurobraille displays, with support for automatic and manual port selection. Note that for Iris displays, port selection should always occur manually, as these are serial displays.

For Esytime, the driver supports switching between protocol based input (i.e. input handled by NVDA) and HID simulation input. Support for HID simulation might be added to Esys in the future.

The driver also contains key bindings as requested by Eurobraille. There ought to be proper support for several firmware versions, including newer braille displays which send acknowledgement packets. Due to the acknowledgement handling required, this driver requires #7732, and therefore this pr contains commits from that pr. All changes to braille.py are already incubating and thus don't need additional review.

Eurobraille wants us to provide other applications access to a device while NVDA is using it. Because of that, hwIo.Hid now takes an additional parameter exclusive, which defaults to True. If set to False as in the Eurobraille driver, access to the device handle is not restricted for other applications. As this is a specific manufacturer request, Any issues caused by this behaviour should be discussed with Eurobraille.

### Testing performed:
I tested an Esys 12 and Esys 40 myself. I also owe many thanks to @clementb49 and @Andre9642, the former tested with an Esytime Evo I belief. Last but not least, many thanks go out to Eurobraille, as they performed tests with Esys 64 and 80, Esytime and Iris displays.

### Known issues with pull request:
Eurobraille requested some key assignments which could not yet be made (e.g. toggling control or shift). This will be possible whenever #7843 is implemented.

### Change log entry:
* New Features
    + Support for Eurobraille Esys, Esytime and Iris braille displays. (#7488)

* Changes for developers
    + hwIo.Hid now takes an additional parameter exclusive, which defaults to True. If set to False, other applications are allowed to communicate with a device while it is connected to NVDA. (#7859)

### CODE REVIEW

File: source/winKernel.py [hunk 1]
```diff
@@ -55,6 +56,58 @@ def CreateFile(fileName,desiredAccess,shareMode,securityAttributes,creationDispo
 	return res
 
 
+def createWaitableTimer(securityAttributes=None, manualReset=False, name=None):
```
Comment (michaelDCurran) [inline source/winKernel.py]: I think github is not showing this right... but it looks like CreateWaitableTimer and SetWaitableTimer are being added here, yet they already exist in master I think.

**Review by michaelDCurran — COMMENTED** [review]
@leonardder: could you please merge master into this pr again now that acknowledgement packets has been merged to master? This will make the diff easier to read. Strangely  git diff and github are still showing a lot of code being added which is already in master, such as CreateWaitableTimer.

**Comment by michaelDCurran** [comment]
@leonardder: could you please merge master into this pr again now that acknowledgement packets has been merged to master? This will make the diff easier to read. Strangely  git diff and github are still showing a lot of code being added which is already in master, such as CreateWaitableTimer.

**Comment by leonardder** [comment]
Just did :)

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -407,8 +407,8 @@ Review modes are a replacement for the older Flat Review concept found in NVDA.
 The following commands switch between review modes:
 %kc:beginInclude
 || Name | Desktop key | Laptop key | Touch | Description |
-| switch to next review mode | NVDA+numpad7 | NVDA+pageUp | 2-finger flick up | switches to the next available review mode |
-| switch to previous review mode | NVDA+numpad1 | NVDA+pageDown | 2-finger flick down | switches to the previous available review mode |
+| Switch to next review mode | NVDA+numpad7 | NVDA+pageUp | 2-finger flick up | switches to the next available review mode |
+| Switch to previous review mode | NVDA+numpad1 | NVDA+pageDown | 2-finger flick down | switches to the previous available review mode |
```
[the remark this answers is not in the document]
Comment (leonardder) [reply user_docs/en/userGuide.t2t]: The initial command descriptions did not start with an upper case letter. Now they do. I actually had forgotten I had changed this. I can revert if you like, but it looked more consistent.

File: source/winKernel.py [hunk 3]
```diff
@@ -46,6 +46,7 @@ def GetStdHandle(handleID):
 GENERIC_WRITE=0x40000000
 FILE_SHARE_READ=1
 FILE_SHARE_WRITE=2
+FILE_SHARE_DELETE=4
```
Comment (leonardder) [inline source/winKernel.py]: This one isn't used, but it was the only access right constant that was missing, so people who use winKernel might expect it should be there.

**Review by michaelDCurran — APPROVED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**